common/hashing.h

```diff
@@ -0,0 +1,843 @@
+// a free function that can be found by ADL for your type -- called `CarbonHash`
+// with the following signature:
+//
+// ```cpp
+// auto CarbonHash(const YourType& value, uint64_t seed) -> HashCode;
+// ```
+//
+// This function needs to ensure that values that compare equal (including any
+// comparisons with different types that might be used with a hash table of
+// `YourType` keys) produce the same `HashCode` values. Input values that
+// compare unequal should work to produce `HashCode` values with as many
+// different bits as possible.
+//
+// `HashCode` values should typically be produced using the `Hasher` helper type
+// below. See its documentation for more details about implementing these
+// customization points and how best to incorporate the value's state into a
+// `HashCode`.
+template <typename T>
+inline auto HashValue(const T& value, uint64_t seed) -> HashCode;
+
+// The same as the seeded version of `HashValue` but without callers needing to
+// provide a seed.
+//
+// Generally prefer the seeded version, but this is available if there is no
+// reasonable seed. In particular, this will behave better than using a seed of
+// `0`. One important use case is for recursive hashing of sub-objects where
+// appropriate or needed.
+template <typename T>
+inline auto HashValue(const T& value) -> HashCode;
+
+// Object and APIs that eventually produce a hash code.
+//
+// This type is primarily used by types to implement a customization point
+// `CarbonHash` that will in turn be used by the `HashValue` function. See the
+// `HashValue` function for details of that extension point.
+//
+// The methods on this type can be used to incorporate data from your
+// user-defined type into its internal state which can be converted to a
+// `HashCode` at any time. These methods will only produce the same `HashCode`
+// if they are called in the exact same order with the same arguments -- there
+// are no guaranteed equivalences between calling different methods.
+//
+// Example usage:
+// ```cpp
+// auto CarbonHash(const MyType& value, uint64_t seed) -> HashCode {
+//   Hasher hasher(seed);
+//   hasher.HashTwo(value.x, value.y);
+//   return static_cast<HashCode>(hasher);
+// }
+// ```
+//
+// Another important note is that types only need to include the data that would
+// lead to an object comparing equal or not-equal to some other object,
+// including objects of other types if you intend to support heterogeneous
+// hash table lookups between those types.
+//
+// To illustrate this -- if a type has some fixed amount of data, maybe 32
+// 4-byte integers, and equality comparison only operates on the *values* of
+// those integers, then the size (32 integers, or 128 bytes) doesn't need to be
+// included in the hash. But if a type has a *dynamic* amount of data, and the
+// sizes are compared as part of equality comparison, then that dynamic size
+// should typically be included in the hash.
```

Comment:
I don't think they ignore the size, they handle it implicitly: treating end-of-iteration as not-equal to any other value is a way of comparing the size.

And the analogous approach applies equally well here. I actually had an example illustrating this but I think one of the suggestions was to remove it: you can equally hash some sentinel as "end-of-iteration" (provided it is different from any other value), and that works. You might for example hash the null byte in a C-string rather than computing the length up front.

Not sure what text to use to clarify this, as my previous attempt seemed to clearly not work given the suggestion to remove it...

Reply:
Let me try approaching this from the other direction: does anything actually go wrong if I don't include the dynamic size in the hash computation? It certainly doesn't lead to equal values having unequal hashes, and it's hard for me to see how it would lead to meaningfully lower hash quality, especially given the relaxed quality standards we have here.

Reply:
I think what I'm getting stuck on is that it's hard to see how those problems connect to the question of whether the equality comparison compares the sizes. For example, I think it's reasonable to argue that "treating end-of-iteration as not-equal to any other value is a way of comparing the size", but I don't think readers will necessarily see it that way.

I'm not sure how to fix that, but maybe we can cut the Gordian knot by just dropping this passage? We don't need to tell users that `HashSizedBytes` has to incorporate the size, because `HashSizedBytes` can just do that, so this passage is only really relevant to them if their hash function is manually looping over their data. But given what you've said elsewhere about how when users care about hash quality/performance they should explicitly structure the data to enable that, maybe we don't need to worry so much about hash quality in the manual-loop case?

If we do need to care about hash quality in the manual-loop case, maybe we could instead say that in some cases, certain `Hasher` methods may be idempotent, so if they're being called a variable number of times, the number of iterations should be explicitly mixed into the hash value to avoid collisions?